File: ViewBase/Source/ViewModel/BaseViewModel.vb
Imports System.ComponentModel
Imports System.Runtime.CompilerServices
Imports System.Windows
Imports System.Windows.Input
Imports FluentValidation
Imports FluentValidation.Results
Imports Wells.Base

Public MustInherit Class BaseViewModel
    Implements INotifyPropertyChanged, IValidatorProvider

    Protected validators As New Dictionary(Of String, IValidator)
    Protected failures As New Dictionary(Of String, ValidationResult)
    Protected validationResults As New Dictionary(Of String, ValidationResult)
    Protected commandNotifications As New Dictionary(Of String, IEnumerable(Of ICommand))

    Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

    Protected Sub NotifyPropertyChanged(propertyName As String)
        RaiseCommandUpdates(propertyName)
        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
    End Sub

    Protected Sub ShowErrorMessage(message As String)
        View?.ShowErrorMessageBox(message)
    End Sub

    Protected Sub ShowMessage(message As String, title As String)
        View?.ShowMessageBox(message, title)
    End Sub

    Protected Sub OnError(ex As Exception)
        Application.Current.Dispatcher.Invoke(Sub()
                                                  ExceptionHandler.Handle(ex, False)
                                                  Dim errorMessage = ExceptionHandler.GetAllMessages(ex).Trim
                                                  ShowErrorMessage(errorMessage)
                                              End Sub)
    End Sub

    Protected Overridable Property View As IView

    Sub New(view As IView)
        If view IsNot Nothing Then
            SetView(view)
        End If
        _progress = New Progress(Of Integer)(AddressOf ProgressChanged)
    End Sub

    Sub SetView(view As IView)
        Me.View = view
        NotifyPropertyChanged(NameOf(CloseModalViewCommand))
        NotifyPropertyChanged(NameOf(CloseNonModalViewCommand))
        OnSetView(view)
    End Sub

    Protected Overridable Sub OnSetView(view As IView)

    End Sub

    ReadOnly Property Errors As String
        Get
            Return String.Join(vbCrLf, From v In validationResults.Values
                                       Where Not v.IsValid
                                       Select String.Join(vbCrLf, v.Errors.Select(Function(e) e.ErrorMessage)))
        End Get
    End Property

    ReadOnly Property HasFailures As Boolean
        Get
            If validators.Any AndAlso validationResults.Count <> validators.Count Then
                Return True
            End If
            Return validationResults.Values.Any(Function(r) Not r.IsValid)
        End Get
    End Property

    Protected Overridable Sub Initialize()
        SetValidators()
        SetCommandUpdates()
    End Sub

    Protected MustOverride Sub SetValidators()

    Protected MustOverride Sub SetCommandUpdates()

    Protected Sub ValidateAll()
        For Each key In validators.Keys
            Validate(key, CallByName(Me, key, CallType.Get))
        Next
    End Sub

    Protected Sub Validate(Of T)(propertyName As String, value As T)
        validationResults(propertyName) = validators(propertyName).Validate(value)
        If IsValid(propertyName) Then
            OnPropertyValidated(propertyName)
        End If
    End Sub

    Sub AddValidator(propertyName As String, validator As IValidator)
        validators.Add(propertyName, validator)
    End Sub

    Sub AddCommands(propertyName As String, commands As IEnumerable(Of ICommand))
        If commandNotifications.ContainsKey(propertyName) Then
            Dim newlist = commandNotifications(propertyName).ToList
            newlist.AddRange(commands)
            commandNotifications(propertyName) = newlist
        Else
            commandNotifications.Add(propertyName, commands)
        End If
    End Sub

    Protected Sub SetValue(Of T)(ByRef value As T, newValue As T, <CallerMemberName()> Optional propertyName As String = Nothing, Optional forceAssignment As Boolean = False)
        If (newValue Is Nothing And value IsNot Nothing) OrElse (newValue IsNot Nothing AndAlso (Not newValue.Equals(value) Or forceAssignment)) Then
            OnPropertyChanging(propertyName)
            value = newValue
            If validators.ContainsKey(propertyName) Then
                Validate(propertyName, newValue)
            End If
            NotifyPropertyChanged(propertyName)
            OnPropertyChanged(propertyName)
            UpdateCommandsThatDependsOnFailures()
        End If
    End Sub

    Protected Overridable Sub OnPropertyChanging(propertyName As String)

    End Sub

    Protected Overridable Sub OnPropertyChanged(propertyName As String)

    End Sub

    Protected Overridable Sub OnPropertyValidated(propertyName As String)

    End Sub

    Protected Overridable Sub RaiseCommandUpdates(propertyName As String)
        If commandNotifications.ContainsKey(propertyName) Then
            For Each com In commandNotifications(propertyName)
                CType(com, RelayCommand).RaiseCanExecuteChanged()
            Next
        End If
    End Sub

    Protected Sub UpdateCommandsThatDependsOnFailures()
        NotifyPropertyChanged(NameOf(Errors))
        NotifyPropertyChanged(NameOf(HasFailures))
        RaiseCommandUpdates(NameOf(Errors))
        RaiseCommandUpdates(NameOf(HasFailures))
    End Sub

    Function ValidationResult(propertyName As String) As ValidationResult Implements IValidatorProvider.ValidationResult
        Dim result As ValidationResult = Nothing
        If propertyName Is Nothing OrElse Not validationResults.TryGetValue(propertyName, result) Then
            Return New ValidationResult()
        Else
            Return result
        End If
    End Function

    Function IsValid(propertyName As String) As Boolean Implements IValidatorProvider.IsValid
        If propertyName Is Nothing Then Return True
        Return ValidationResult(propertyName).IsValid
    End Function

    Property CloseModalViewCommand As ICommand = New RelayCommand(Sub(param)
                                                                      Dim result As Boolean?
                                                                      If param Is Nothing Then
                                                                          result = False
                                                                      Else
                                                                          result = CType(param, Boolean?)
                                                                      End If
                                                                      View?.CloseView(result)
                                                                  End Sub, Function()
                                                                               Return View IsNot Nothing
                                                                           End Function, AddressOf OnError)

    Property CloseNonModalViewCommand As ICommand = New RelayCommand(Sub()
                                                                         View?.CloseView()
                                                                     End Sub, Function()
                                                                                  Return View IsNot Nothing
                                                                              End Function, AddressOf OnError)


    Protected WithEvents _progress As IProgress(Of Integer)
    Protected _processInfo As String
    ReadOnly Property ProcessInfo As String
        Get
            Return _processInfo
        End Get
    End Property

    Protected _showProgress As Boolean
    ReadOnly Property ShowProgress As Boolean
        Get
            Return _showProgress
        End Get
    End Property

    Protected _undefinedProgress As Boolean
    ReadOnly Property UndefinedProgress As Boolean
        Get
            Return _undefinedProgress
        End Get
    End Property

    Protected _progressValue As Integer
    ReadOnly Property ProgressValue As Integer
        Get
            Return _progressValue
        End Get
    End Property

    Protected Sub ProgressChanged(value As Integer)
        _progressValue = value
        NotifyPropertyChanged(NameOf(ProgressValue))
    End Sub

    Protected Sub StartProgressNotifications(undefined As Boolean, Optional message As String = "")
        _progressValue = 0
        _undefinedProgress = undefined
        _processInfo = message
        _showProgress = True
        NotifyPropertyChanged(NameOf(ProgressValue))
        NotifyPropertyChanged(NameOf(ProcessInfo))
        NotifyPropertyChanged(NameOf(UndefinedProgress))
        NotifyPropertyChanged(NameOf(ShowProgress))
    End Sub

    Protected Sub StopProgressNotifications()
        _undefinedProgress = False
        _processInfo = String.Empty
        _showProgress = False
        NotifyPropertyChanged(NameOf(ProcessInfo))
        NotifyPropertyChanged(NameOf(UndefinedProgress))
        NotifyPropertyChanged(NameOf(ShowProgress))
    End Sub

End Class

File: BaseView/src/ViewModel/BaseViewModel.cs
using FluentValidation;
using FluentValidation.Results;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Windows;
using System.Windows.Input;
using Wells.Base;
using Wells.BaseView.ViewInterfaces;

namespace Wells.BaseView.ViewModel
{
    public abstract class BaseViewModel : INotifyPropertyChanged
    {
        protected Dictionary<string, IEnumerable<IValidator>> validators = new Dictionary<string, IEnumerable<IValidator>>();
        protected Dictionary<string, List<ValidationResult>> failures = new Dictionary<string, List<ValidationResult>>();
        protected Dictionary<string, List<ValidationResult>> validationResults = new Dictionary<string, List<ValidationResult>>();
        protected Dictionary<string, IEnumerable<ICommand>> commands = new Dictionary<string, IEnumerable<ICommand>>();
        protected virtual IView View { get; set; }

        public event PropertyChangedEventHandler PropertyChanged;


        protected BaseViewModel(IView view)
        {
            if (view != null)
            {
                SetView(view);
            }
        }

        public void SetView(IView view)
        {
            View = view;
            NotifyPropertyChanged(nameof(CloseModalViewCommand));
            NotifyPropertyChanged(nameof(CloseNonModalViewCommand));
            OnSetView(view);
        }

        protected virtual void OnSetView(IView view) { }

        public bool IsValid(string propertyName)
        {
            if (string.IsNullOrEmpty(propertyName)) { return true; }
            return ValidationResult(propertyName).All(r => r.IsValid);
        }

        public List<ValidationResult> ValidationResult(string propertyName)
        {
            if (propertyName != null || validationResults.ContainsKey(propertyName))
            { return validationResults[propertyName]; }
            return new List<ValidationResult>();
        }

        protected void NotifyPropertyChanged(string propertyName)
        {
            RaiseCommandUpdates(propertyName);
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }

        protected void ShowErrorMessage(string message)
        {
            if (View != null)            
            { 
                SharedBaseView. ShowErrorMessageBox(View, message); 
            }
        }

        protected void ShowMessage(string message, string title)
        {
            if (View != null)
            {
                SharedBaseView.ShowOkOnkyMessageBox(View, message, title);
            }
        }

        protected bool ShowYesNoMessageBox(string message, string title)
        {
            if (View != null) { return SharedBaseView.ShowYesNoMessageBox(View, message, title); }
            return false;
        }

        protected void OnError(Exception ex)
        {
            Application.Current.Dispatcher.Invoke(() =>
            {
                ExceptionHandler.Handle(ex, false);
                var errorMessage = ExceptionHandler.GetAllMessages(ex).Trim();
                ShowErrorMessage(errorMessage);
            });
        }

        protected void SetValue<T>(ref T value, T newValue, [CallerMemberName()] string propertyName = null, bool forceAssignment = false)
        {
            if ((newValue == null && value != null) || (newValue != null && (!newValue.Equals(value) || forceAssignment)))
            {
                OnPropertyChanging(propertyName);
                value = newValue;
                if (validators.ContainsKey(propertyName)) { Validate(propertyName, newValue); }
                NotifyPropertyChanged(propertyName);
                OnPropertyChanged(propertyName);
                UpdateCommandsThatDependsOnFailures();
            }
        }

        protected virtual void OnPropertyChanging(string propertyName) { }

        protected virtual void OnPropertyChanged(string propertyName) { }


        protected virtual void OnPropertyValidated(string propertyName) { }


        protected virtual void RaiseCommandUpdates(string propertyName)
        {
            if (commands.ContainsKey(propertyName))
            {
                foreach (ICustomCommand command in commands[propertyName])
                {
                    command.OnCanExecuteChanged();
                }
            }
        }

        protected abstract void SetValidators();

        protected abstract void SetCommandUpdates();

        protected void Add(string propertyName, IEnumerable<IValidator> validators)
        {
            if (this.validators.ContainsKey(propertyName))
            {
                var list = this.validators[propertyName].ToList();
                list.AddRange(validators);
                this.validators[propertyName] = list;
            }
            else
            { this.validators.Add(propertyName, validators); }
        }

        protected void Add(string propertyName, IValidator validator)
        {
            if (validators.ContainsKey(propertyName))
            {
                var list = validators[propertyName].ToList();
                list.Add(validator);
                this.validators[propertyName] = list;
            }
            else
            { this.validators.Add(propertyName, new List<IValidator>() { validator }); }
        }

        protected void Add(string propertyName, IEnumerable<ICommand> commands)
        {
            if (this.commands.ContainsKey(propertyName))
            {
                var list = this.commands[propertyName].ToList();
                list.AddRange(commands);
                this.commands[propertyName] = list;
            }
            else
            { this.commands.Add(propertyName, commands); }
        }

        protected void Add(string propertyName, ICommand command)
        {
            if (commands.ContainsKey(propertyName))
            {
                var list = commands[propertyName].ToList();
                list.Add(command);
                commands[propertyName] = list;
            }
            else
            { commands.Add(propertyName, new List<ICommand>() { command }); }
        }

        protected virtual void Initialize()
        {
            SetValidators();
            SetCommandUpdates();
        }

        protected void UpdateCommandsThatDependsOnFailures()
        {
            NotifyPropertyChanged(nameof(Errors));
            NotifyPropertyChanged(nameof(HasFailures));
            RaiseCommandUpdates(nameof(Errors));
            RaiseCommandUpdates(nameof(HasFailures));

        }

        protected void ValidateAll()
        {
            foreach (var key in validators.Keys)
            {
                Validate(key, Microsoft.VisualBasic.Interaction.CallByName(this, key, Microsoft.VisualBasic.CallType.Get));
            }
        }


        protected void Validate<T>(string propertyName, T value)
        {
            if (value == null)
            {
                var errors = new List<ValidationFailure>() { new ValidationFailure(propertyName, "Error: hay una propiedad nula") };
                validationResults[propertyName] = new List<ValidationResult> { new ValidationResult(errors) };
            }
            else
            {
                var results = new List<ValidationResult>();
                foreach (var v in validators[propertyName])
                {
                    results.Add(v.Validate(value));
                }

                validationResults[propertyName] = results;
            }
            if (IsValid(propertyName)) { OnPropertyValidated(propertyName); }
        }

        public string Errors
        {
            get
            {
                var results = new List<ValidationResult>();
                foreach (var r in validationResults.Values)
                {
                    results.AddRange(r);
                }

                return string.Join(Environment.NewLine, from v in results
                                                        where !v.IsValid
                                                        select string.Join(Environment.NewLine, v.Errors.Select(e => e.ErrorMessage)));
            }
        }

        public bool HasFailures
        {
            get
            {
                if (validators.Any() && validationResults.Count != validators.Count) { return true; }
                return validationResults.Values.Any(r => r.Any(vr => !vr.IsValid));
            }
        }

        public ICommand CloseModalViewCommand
        {
            get => new RelayCommand((param) =>
            {
                bool? result;
                if (param == null) { result = false; }
                else { result = (bool?)param; }
                View?.CloseView(result);
            }, (obj) => View != null, OnError);
        }


        public ICommand CloseNonModalViewCommand
        {
            get => new RelayCommand((param) =>
            {
                View?.CloseView();
            }, (obj) => View != null, OnError);
        }

    }
}
